lib/charms/opensearch/v0/opensearch_locking.py

```diff
@@ -25,78 +25,242 @@
+    def _unit_with_lock(self, value: str):
+        assert self._relation
+        self._relation.data[self._charm.app]["unit-with-lock"] = value
+
+    @_unit_with_lock.deleter
+    def _unit_with_lock(self):
+        assert self._relation
+        self._relation.data[self._charm.app].pop("unit-with-lock", None)
+
+    @property
+    def _relation(self):
+        # Use property instead of `self._relation =` in `__init__()` because of ops Harness unit
+        # tests
+        return self._charm.model.get_relation(self._ENDPOINT_NAME)
+
+    def _on_peer_relation_changed(self, _=None):
+        """Grant & release lock."""
+        if not self._charm.unit.is_leader():
+            return
+        assert self._relation
+        if self._unit_with_lock and self._unit_requested_lock(
+            self._charm.model.get_unit(self._unit_with_lock)
+        ):
+            # Lock still in use, do not release
+            logger.debug("[Node lock] (leader) lock still in use")
+            return
+        # TODO: adjust which unit gets priority on lock?
+        for unit in (*self._relation.units, self._charm.unit):
+            if self._unit_requested_lock(unit):
+                self._unit_with_lock = unit.name
+                logger.debug(f"[Node lock] (leader) granted peer lock to {unit.name=}")
+                break
+        else:
+            logger.debug("[Node lock] (leader) cleared peer lock")
+            del self._unit_with_lock
+
+
+class OpenSearchNodeLock(ops.Object):
+    """Ensure that only one node (re)starts, joins the cluster, or leaves the cluster at a time.
 
-        # we check first on the peer data bag if the lock is already acquired
-        if self._is_lock_in_peer_data():
-            raise OpenSearchOpsLockAlreadyAcquiredError("Another unit is being removed.")
+    Uses OpenSearch document for lock. Falls back to peer databag if no units online
+    """
 
-        host = self._charm.unit_ip if self._opensearch.is_node_up() else None
+    _OPENSEARCH_INDEX = ".charm_node_lock"
 
-        # we can use opensearch to lock
-        if host is not None or self._charm.alt_hosts:
+    def __init__(self, charm: "opensearch_base_charm.OpenSearchBaseCharm"):
+        super().__init__(charm, "opensearch-node-lock")
+        self._charm = charm
+        self._opensearch = charm.opensearch
+        self._peer = _PeerRelationLock(self._charm)
+
+    def _unit_with_lock(self, host) -> str | None:
+        """Unit that has acquired OpenSearch lock."""
+        try:
+            document_data = self._opensearch.request(
+                "GET",
+                endpoint=f"/{self._OPENSEARCH_INDEX}/_source/0",
+                host=host,
+                alt_hosts=self._charm.alt_hosts,
+                retries=3,
+            )
+        except OpenSearchHttpError as e:
+            if e.response_code == 404:
+                # No unit has lock
+                return
+            raise
+        return document_data["unit-name"]
```

Comment:
Actually, it would be better that only the charm app leader that is the "cluster_manager" to do this check, and it would check against all the units related to it via peer && large deployment clusters. This way we do not delete that entry for no reason.